### ISSUE TITLE
Bedrock Upgrade Process

### ISSUE THREAD

**Issue by wsdt** [bobanetwork/v3-boba#7]
### Upgrade Process
The high level upgrade process looks like this:

[#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#optimism-upgrade-steps)Optimism Upgrade Steps
1. At a predefined block height, we disable the sequencer and take a state snapshot.
2. We upgrade our smart contracts.
3. We generate a genesis.json file from the state snapshot.
4. We initialize the sequencer with the genesis.json file.
5. We redeploy our Legacy Geth instances with the state snapshot.
6. We give infra providers the following artifacts:
- The genesis.json file for the sequencer.
- The state snapshot for Legacy Geth.
- The genesis hashes that the Rollup Node needs as part of its configuration.
7. We wait for infrastructure providers to upgrade.
8. We re-enable the sequencer.
9. Now the chain is bedrock enabled!

[#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#infra-provider-upgrade-steps)Infra Provider Upgrade Steps
Upon receipt of the genesis.json, state snapshot, and genesis hashes:

1. Redeploy Legacy Geth with the provided state snapshot.
2. Initialize Bedrock Geth with the provided genesis.json file.
3. Configure the Rollup Node to use the provided genesis hashes.

Source: https://community.optimism.io/docs/developers/bedrock-temp/infra/#upgrade-process

FYI: 
@InoMurko @mmontour1306 @Boyuan-Chen

**Comment by wsdt** [comment]
How to run a node on their Alpha testnet: 
https://oplabs.notion.site/Running-a-Node-eda545c730e64b44b762ab12e93296aa

**Comment by wsdt** [comment]
### Guidelines
1. The Bedrock Geth node is deployed identically to an L1 Geth node:
- It needs access to Ethereum’s peer-to-peer network.
- It needs a large attached disk to store its state.
- Its RPC ports need to be accessible in order for users to submit transactions and read chain data.
- The Bedrock Geth node needs to be accessible by the Rollup Node in order for the Engine API to function correctly.
2. The Rollup Node can be deployed as a single container.
- It needs access to an L1 RPC endpoint.
- It needs access to the Bedrock Geth node, preferably over a WebSocket connection.
- It needs access to the Rollup Node’s peer-to-peer network.
- The Rollup Node is stateless, so it does not need a writable disk.
3. Don’t try to run multiple Bedrock Geth nodes per Rollup Node or vice versa. Always deploy them in pairs.
- In a Kubernetes deployment, the Rollup Node might be deployed as a sidecar to the Bedrock Geth node.
4. The Legacy Geth node is deployed similarly to an L1 Geth node, except:
- The node is read-only, so it will not run any transaction pooling or sync logic.
- Since the state of the Legacy Geth node is immutable, it is safe to run unlimited Legacy Geth nodes behind a load balancer without worrying about consistency problems.
5. The Daisy Chain is deployed similarly to any proxy container:
- It needs access to the RPC endpoints of a Legacy Geth and Bedrock Geth node.
- The Geth nodes the Daisy Chain routes to can be behind load balancers, as long as the load balancers only contain Geth nodes of a single type (i.e., do not mix Legacy and Bedrock Geth nodes behind the same LB).
[#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#upgrade-plan)

**Comment by wsdt** [comment]
Purpose of legacy Geth: 

### Legacy Geth
- A stripped-down, read-only version of the legacy (i.e., pre-Bedrock) l2geth node.
- Serves historical data to the Daisy Chain.
- The Legacy Geth instance will ship with a pre-populated database. It does not need state synchronization, because that database is real only.
[#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#daisy-chain)

**Comment by InoMurko** [comment]
Awesome, thank you.

**Comment by boyuan-chen** [comment]
How do they handle the block height of the new node? If the node starts from 0 block, developers need to fix their smart contracts due to the changes of the block height.

**Comment by InoMurko** [comment]
Block height in Bedrock is derived from L1, it's not independent of block production on L2, like it was pre-Bedrock.

**Comment by boyuan-chen** [comment]
So the block number changes from the pre-bedrock to bedrock. Developers have to be aware of the potential issues and fix that in the contracts and we apply the new bytecode to the state dump file.

**Comment by InoMurko** [comment]
> So the block number changes from the pre-bedrock to bedrock. Developers have to be aware of the potential issues and fix that in the contracts and we apply the new bytecode to the state dump file.

Yeah, if their dapp would involve the assumption of block continuuance of N+1, this could possibly cause issues. We need to be aware of that and put the message out and provide support. Agreed, thanks for noticing this.

**Comment by InoMurko** [comment]
Oh actually - I'm wrong:
```
BLOCKNUMBER will advance every two seconds because we'll create a new block every two seconds, regardless of the number of transactions.
[#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#contracts)
```


https://community.optimism.io/docs/developers/bedrock/#blocks
https://community.optimism.io/docs/developers/bedrock/#rollup-node

What comes from L1 is the state root, not the block directly!
So I guess the block number it starts from should be in the genesis file?

**Comment by boyuan-chen** [comment]
> Oh actually - I'm wrong:
> 
> ```
> BLOCKNUMBER will advance every two seconds because we'll create a new block every two seconds, regardless of the number of transactions.
> [#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#contracts)
> ```
> 
> https://community.optimism.io/docs/developers/bedrock/#blocks https://community.optimism.io/docs/developers/bedrock/#rollup-node
> 
> What comes from L1 is the state root, not the block directly! So I guess the block number it starts from should be in the genesis file?

I think genesis file doesn't define the block number. My understanding is that it will start from 0 block.

**Comment by InoMurko** [comment]
> > Oh actually - I'm wrong:
> > ```
> > BLOCKNUMBER will advance every two seconds because we'll create a new block every two seconds, regardless of the number of transactions.
> > [#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#contracts)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://community.optimism.io/docs/developers/bedrock/#blocks https://community.optimism.io/docs/developers/bedrock/#rollup-node
> > What comes from L1 is the state root, not the block directly! So I guess the block number it starts from should be in the genesis file?
> 
> I think genesis file doesn't define the block number. My understanding is that it will start from 0 block.

Where is your understanding coming from? 
Could be it starts from 0 tho I think that would break lots of dapps and etherscan etc.

**Comment by boyuan-chen** [comment]
> > > Oh actually - I'm wrong:
> > > ```
> > > BLOCKNUMBER will advance every two seconds because we'll create a new block every two seconds, regardless of the number of transactions.
> > > [#](https://community.optimism.io/docs/developers/bedrock-temp/infra/#contracts)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > https://community.optimism.io/docs/developers/bedrock/#blocks https://community.optimism.io/docs/developers/bedrock/#rollup-node
> > > What comes from L1 is the state root, not the block directly! So I guess the block number it starts from should be in the genesis file?
> > 
> > 
> > I think genesis file doesn't define the block number. My understanding is that it will start from 0 block.
> 
> Where is your understanding coming from? Could be it starts from 0 tho I think that would break lots of dapps and etherscan etc.

Based on the current state dump file that we use and we tried to start the chain from a random x block when we changed from V1 to V2, it actually breaks the geth client if you don't start from 0 block. Also, Optimism mentions that
```
 The last testnet will replace our existing Goerli, and run for an entire month to give developers time to fix any integration issues they may encounter.
```
They have already done this before from v1 to v2.

**Comment by boyuan-chen** [comment]
Of course, that's my guess. We can see if Optimism plans to do this kind of things.

**Comment by wsdt** [comment]
Ping, if we want to keep this open or if it can be closed

**Comment by mmontour1306** [comment]
Let's keep this one open for now, until we come up with a set of new stories in Zenhub (or whatever) to replace it.
